### ISSUE TITLE
Support moving data out of PG11 default partition

### ISSUE THREAD

**Issue by keithf4** [pgpartman/pg_partman#230]
Add support to partition_data*() functions for moving data out of the PG11 default partition into the relevant child table(s).

Will involve using a temporary table to hold data since child table with a constraint covering data in the default cannot be added.

**Comment by Airconaaron** [comment]
Hey any updates on this feature? Would love to help out in some way.

Currently at my company we had an issue where it turns out the run_maintenance() was never run since the beginning, so our default tables are massive.

With this sort of feature I imagine the logic would be similar for resizing the partition sizes as well.

**Comment by keithf4** [comment]
Just a lack of time for the delay on this. Basic feature of this would be that it has to first move the data out to some other table (either temporary or not) before the relevant child table can be created for it. If you'd like to take a crack at adding this feature, I'd be glad to review and help.

**Comment by keithf4** [comment]
This feature has been added to 4.1.0. Please let me know if it works ok for you.

**Comment by mewwts** [comment]
Doesn't seem to work for me. I'm getting the issue in #247 when running `run_maintenance`

`ERROR:  updated partition constraint for default partition would be violated by some row`

**Comment by keithf4** [comment]
Do you have data in your default partition? Moving it out isn't an automatic operation, you will have to run the `partition_data*()` function/procedure to do so.

**Comment by mewwts** [comment]
Yes, I do. I must have set up the background worker incorrectly, because partitions were not made automatically. All the data in my default partition belongs in partitions not created. What happens when I then `CALL partition_data_proc('x')`?

**Comment by keithf4** [comment]
For native partitioning, it moves all the data that would go into a single child partition for the given interval in a single transaction. Then commits and if any data is still in the default, it will move on to creating the next child and commit after each child is created until the default table is empty.

You can review additional options for `partition_data_proc()`  in the docs https://github.com/pgpartman/pg_partman/blob/master/doc/pg_partman.md#creation-functions

I'd recommend reviewing all the options available to see if any may allow your run of the function to run smoother for your environment.

**Comment by mewwts** [comment]
Does it create the necessary child tables and attaches them too?

Thanks, I'm fortunate that I don't have live traffic on this instance yet so I don't mind the locks.

**Comment by keithf4** [comment]
Yes. The procedure basically just calls the relevant `partition_data_time()` or `partition_data_id()` functions under the hood in a loop. So can read up on those for more details on it does.

**Comment by mewwts** [comment]
Thank you @keithf4!

**Comment by nobleread** [comment]
I've experienced a problem while calling **partition_data_proc** for time-based partitioned table under partman's control (partitioned natively). Issue was that after adding a single future record, i.e. a record out of retention/premake range, to a (default) table and calling this procedure, I got stuck in a never ending loop in **partition_data_time** function after `ELSIF v_partition_type IN ('time-custom', 'native')` statement - loop's exit condition could never be satisfied and only changing `IF v_start_control > v_max_partition_timestamp THEN` to `IF v_start_control >= v_max_partition_timestamp THEN`, so changing comparison from `>` to `>=`, allowed for whole procedure to exit and do its job. Is this a mistake, that should be fixed in partman's source code?

Second thing I've noticed is that after this record was moved from default table to a proper partition, it seems like run_maintenance does not work properly any longer after part_config.premake param for this table was changed.

Please take a look how I got both errors below.


So I start off with the table configured like this:
```
-[ RECORD 1 ]--------------+--------------------------------------
parent_table               | public.foobar
control                    | datecolumn
partition_type             | native
partition_interval         | 1 day
constraint_cols            |
premake                    | 14
optimize_trigger           | 4
optimize_constraint        | 30
epoch                      | none
inherit_fk                 | t
retention                  | 37 days
retention_schema           |
retention_keep_table       | f
retention_keep_index       | f
infinite_time_partitions   | t
datetime_string            | YYYY_MM_DD
automatic_maintenance      | on
jobmon                     | f
sub_partition_set_full     | f
undo_in_progress           | f
trigger_exception_handling | f
upsert                     |
trigger_return_null        | t
template_table             | partman.template_public_foobar
publications               |
inherit_privileges         | f
constraint_valid           | t
```

Current time is:
```
db=# select now();               
              now                
-------------------------------- 
 09/26/2019 08:29:57.510308 EDT  
(1 row)                     
```     

Adding a 'future' record:
`db=# insert into foobar values (..., '11/27/2019');
`

And confirm record was put into default table
```
db=# select * from foobar_default;
... | datecolumn
...-+-----------
... | 11/27/2019

```
Call partition_data_proc with a proposed quickfix - everything works and new partition is created, record is moved properly. Without quickfix, I'm stuck in a never ending loop
```
db=# \d+ foobar
...
foobar_p2019_10_09 FOR VALUES FROM ('10/09/2019') TO ('10/10/2019'),
foobar_p2019_10_10 FOR VALUES FROM ('10/10/2019') TO ('10/11/2019'),
foobar_p2019_11_27 FOR VALUES FROM ('11/27/2019') TO ('11/28/2019'),
foobar_default DEFAULT                                              

db=# select count(*) from foobar_default;
 count                                            
-------                                           
     0                                            
(1 row)                                           
```
Set new premake. 
`db=# update partman.part_config set premake = 100;
`

After running partman.run_maintenance, new partitions are created only after the previously last partition and there's a gap between previously last one and one-before last partition, although  infinite_time_partitions is set to true - so every partition should exist
```
...
pcsa_bookmark_p2019_10_09 FOR VALUES FROM ('10/09/2019') TO ('10/10/2019'),
pcsa_bookmark_p2019_10_10 FOR VALUES FROM ('10/10/2019') TO ('10/11/2019'),
pcsa_bookmark_p2019_11_27 FOR VALUES FROM ('11/27/2019') TO ('11/28/2019'),
pcsa_bookmark_p2019_11_28 FOR VALUES FROM ('11/28/2019') TO ('11/29/2019'),
...
pcsa_bookmark_p2020_01_04 FOR VALUES FROM ('01/04/2020') TO ('01/05/2020'),
pcsa_bookmark_p2020_01_05 FOR VALUES FROM ('01/05/2020') TO ('01/06/2020'),
pcsa_bookmark_p2020_01_06 FOR VALUES FROM ('01/06/2020') TO ('01/07/2020'),
pcsa_bookmark_default DEFAULT                  

```
In case I change premake to other value and run partman.run_maintenance on a table where no record was moved from default, all required partitions are created, so this error appears only when non-continuous partitions exist, which is a result of e.g. calling procedure partition_data_proc. In the end I think that premake don't need to be changed in order for that error to appear; it's just that run_maintenance does not properly handle discontinuous partition range.

**Comment by keithf4** [comment]
Would you mind making a new issue for this vs posting to a closed, generalized feature request issue? So I can better keep track of investigating a possible bug. Thank you.

**Comment by nobleread** [comment]
Sure, sorry for that. I've started new issue here:
https://github.com/pgpartman/pg_partman/issues/273
